### PR TITLE
Fixed #24391 -- Call initial callable only once for .value()

### PR DESCRIPTION
**Pull request by orokusaki**
@timgraham - this addresses https://code.djangoproject.com/ticket/24391 without changing the API. 

I've tested this manually, locally.

### CODE REVIEW

**Comment by orokusaki** [comment]
I initially wrote this to accommodate multiple accesses to the field to support specific kinds of manual form usage in a template, etc. (TIL: every time you `form['field']` you get a new instance of `BoundField`). I did this by caching the attribute as `self.form._<field name>_initial_result` Instead of caching on the `BoundField`. I removed that to keep the patch simple at first, in case there needed to be more conversation about this approach in general.

File: django/forms/forms.py [hunk 1]
```diff
@@ -580,7 +580,13 @@ def value(self):
         if not self.form.is_bound:
             data = self.form.initial.get(self.name, self.field.initial)
             if callable(data):
-                data = data()
+                # Get/set the result of a prior call from/to an attribute, to
+                # avoid having to call the initial callable twice (#24391);
+                # default to the result of calling it
```
Comment (timgraham) [inline django/forms/forms.py]: This comment seems a bit wordy. Could we simply say something like `Callables should be evaluated once per field.`? Maybe `self._initial_value` would be a slightly better variable name.

File: tests/forms_tests/tests/test_forms.py [hunk 2]
```diff
@@ -1369,6 +1369,22 @@ class UserRegistration(Form):
         self.assertEqual(bound['password'].value(), 'foo')
         self.assertEqual(unbound['password'].value(), None)
 
+    def test_boundfield_initial_called_once(self):
+        # Ensures multiple calls to BoundField().value() in an unbound form
+        # only results in a single call to a provided initial callable (#24391)
+        initial = mock.Mock(return_value='dynamic')
+
+        class Element(Form):
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: Element -> MyForm

File: tests/forms_tests/tests/test_forms.py [hunk 3]
```diff
@@ -1369,6 +1369,22 @@ class UserRegistration(Form):
         self.assertEqual(bound['password'].value(), 'foo')
         self.assertEqual(unbound['password'].value(), None)
 
+    def test_boundfield_initial_called_once(self):
+        # Ensures multiple calls to BoundField().value() in an unbound form
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: this could probably be a docstring.

File: tests/forms_tests/tests/test_forms.py [hunk 4]
```diff
@@ -1369,6 +1369,22 @@ class UserRegistration(Form):
...
+        # only results in a single call to a provided initial callable (#24391)
+        initial = mock.Mock(return_value='dynamic')
+
+        class Element(Form):
+            name = CharField(max_length=10, initial=initial)
+
+        # Make an instance, and get its "name" value multiple times
+        unbound = Element()
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: form = ...

File: tests/forms_tests/tests/test_forms.py [hunk 5]
```diff
@@ -1369,6 +1369,22 @@ class UserRegistration(Form):
         self.assertEqual(bound['password'].value(), 'foo')
         self.assertEqual(unbound['password'].value(), None)
 
+    def test_boundfield_initial_called_once(self):
+        # Ensures multiple calls to BoundField().value() in an unbound form
+        # only results in a single call to a provided initial callable (#24391)
+        initial = mock.Mock(return_value='dynamic')
+
+        class Element(Form):
+            name = CharField(max_length=10, initial=initial)
```
Comment (timgraham) [inline tests/forms_tests/tests/test_forms.py]: I'm in favor of avoiding mock where possible, so I would use something like `initial=uuid.uuid4` and simply verify the values returned in the two `value()` calls are equal. Are you okay with that approach?

**Comment by orokusaki** [comment]
@timgraham I'll get all those updates in there after work tonight.

In the meantime, what did you think about my comment above, the one pertaining caching the attribute on `self.form`, to allow for many accesses of the same field while only ever calling initial once? Or, as an alternative, do you think it would be feasible to memoize `BoundField` here @ https://github.com/orokusaki/django/blob/master/django/forms/forms.py#L123 (would also be a performance optimization)?

**Comment by timgraham** [comment]
I think it might be better to do the evaluation of `initial` on the form itself. Having the widget evaluate initial and then cache it back on the form seems a bit convoluted. Here is the code I've been playing around with:
```diff
diff --git a/django/forms/fields.py b/django/forms/fields.py
index 6d1c145..eda9825 100644
--- a/django/forms/fields.py
+++ b/django/forms/fields.py
@@ -33,6 +33,7 @@ from django.utils.deprecation import (
 )
 from django.utils.duration import duration_string
 from django.utils.encoding import force_str, force_text, smart_text
+from django.utils.functional import SimpleLazyObject
 from django.utils.ipv6 import clean_ipv6_address
 from django.utils.six.moves.urllib.parse import urlsplit, urlunsplit
 from django.utils.translation import ugettext_lazy as _, ungettext_lazy
@@ -92,7 +93,19 @@ class Field(six.with_metaclass(RenameFieldMethods, object)):
         # localize -- Boolean that specifies if the field should be localized.
         # label_suffix -- Suffix to be added to the label. Overrides
         #                 form's label_suffix.
-        self.required, self.label, self.initial = required, label, initial
+        self.required = required
+        self.label = label
+
+        def lazy_initial():
+            data = initial()
+            # If this is an auto-generated default date, nix the
+            # microseconds for standardized handling. See #22502.
+            if (isinstance(data, (datetime.datetime, datetime.time)) and
+                    not getattr(self.widget, 'supports_microseconds', True)):
+                data = data.replace(microsecond=0)
+            return data
+
+        self.initial = SimpleLazyObject(lazy_initial) if callable(initial) else initial
         self.show_hidden_initial = show_hidden_initial
         self.help_text = help_text
         self.label_suffix = label_suffix
diff --git a/tests/model_formsets/tests.py b/tests/model_formsets/tests.py
index 217a1a1..1fd354f 100644
--- a/tests/model_formsets/tests.py
+++ b/tests/model_formsets/tests.py
@@ -1112,7 +1112,7 @@ class ModelFormsetTest(TestCase):
 
         self.assertEqual(len(formset.forms), 1)  # this formset only has one form
         form = formset.forms[0]
-        now = form.fields['date_joined'].initial()
+        now = form.fields['date_joined'].initial
         result = form.as_p()
         result = re.sub(r'[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]+)?', '__DATETIME__', result)
         self.assertHTMLEqual(result,
@@ -1135,7 +1135,7 @@ class ModelFormsetTest(TestCase):
 
         # now test for when the data changes
 
-        one_day_later = now + datetime.timedelta(days=1)
+        one_day_later = now._wrapped + datetime.timedelta(days=1)
         filled_data = {
             'membership_set-TOTAL_FORMS': '1',
             'membership_set-INITIAL_FORMS': '0',
```
There is a test failure but I'm not sure it represents an actual use case. The model is using this code for the default:
```
callable_default_counter = itertools.count()
callable_default = lambda: next(callable_default_counter)
```
Failure:
```
======================================================================
FAIL: test_formfield_initial (forms_tests.tests.tests.FormsModelTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/forms_tests/tests/tests.py", line 197, in test_formfield_initial
    self.assertNotEqual(r1, r2)
AssertionError: '<input id="id_callable_default" name="callable_default" type="number" value="4" />' == '<input id="id_callable_default" name="callable_default" type="number" value="4" />'======================================================================
FAIL: test_formfield_initial (forms_tests.tests.tests.FormsModelTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/forms_tests/tests/tests.py", line 197, in test_formfield_initial
    self.assertNotEqual(r1, r2)
AssertionError: '<input id="id_callable_default" name="callable_default" type="number" value="4" />' == '<input id="id_callable_default" name="callable_default" type="number" value="4" />'
```

**Comment by orokusaki** [comment]
@timgraham I like your approach much better.

I also strongly agree that having a callable that maintains global (module level), like in the failing test, is probably not a real use case. However, I'm a bit confused as to why that test is failing. It appears as though the form is memoizing the initial result across 2 instantiations of the form. Am I following correctly?

Pertaining the memoizing of `BoundField`, your implementation seems really simple, clean, and straight forward, especially since tests pass.

**Comment by timgraham** [comment]
I am confused by the test failure as well.

**Comment by orokusaki** [comment]
@timgraham Ah, of course. `Field` is only instantiated one time (at a forms module's load time), so `SimpleLazyObject` is being set up with a `_wrapped` value that is the result of calling `initial` the first time it is accessed, which means that `Field` is not a candidate for memoizing the result of `initial`. This means that best viable option might still be to memoize the result on the `BoundField` instance, which requires memoizing the `BoundField` instance on `BaseForm` (your second snippet). If I make a patch with the initial changes you had requested, and with the `bound_fields_cache` code, do you think there's a chance for it to get merged, given proper review from you and some of your peers?

**Comment by MarkusH** [comment]
@timgraham Your second approach (with `self.bound_fields_cache`) seems good. Can we agree on `_bound_field_cache`, though. I don't want people to use or even rely on it.

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/4206

**Comment by MarkusH** [comment]
Closing this PR in favor of #4206 opened by @timgraham . Thank you for the patch, @orokusaki :smile:

**Comment by orokusaki** [comment]
@MarkusH thanks for the heads up. I meant to go check if out right after work, but I'm just now getting to it.
